src/sfloader/fluid_defsfont.c

```diff
@@ -1268,23 +1450,548 @@ fluid_zone_is_mod_identic(fluid_mod_t *mod, char *name)
+                }
+            }
+        }
+        mod = mod->next;
+        mod_idx++;
+    }
+    return ((new_idx == 0) ? total_linked_count : linked_count);
+}
+/**
+ * Checks all modulators from a zone modulator list and optionally clone
+ * valid linked modulators from mod_list list to linked_mod list.
+ * - check valid sources (if requested, see list_name).
+ * - check identic modulators (if requested, see list_name).
+ * - check linked modulators paths.
+ * - check "isolated" linked paths (if requested, see list_name).
+ * - clone valid linked modulators paths to linked_mod.
+ * The function does the same job that fluid_zone_check_mod() except that
+ * that modulators aren't removed from mod_list and lists length aren't
+ * limited. The function is appropriate to be called by API
+ * fluid_voice_add_mod2(),fluid_synth_add_default_mod2().
+ *
+ * @param list_name, list name used to prefix warning messages displayed.
+ *  If NULL, the function does minimum safe check for linked modulators.
+ *   - no warning message are displayed.
+ *   - checks not done are: valid source, identic modulator, isolated linked path.
+ *   This is useful for performance reason, when the caller already know that
+ *   modulators are valid. In this case, the function call only
+ *   fluid_check_linked_mod_path() and fluid_list_copy_linked_mod().
+ *
+ * @param list_mod, pointer on modulators list.
+ *  On input, the list may contains any unlinked or linked modulators.
+ *  On output, invalid modulators are marked invalid with amount value forced
+ *  to 0.
+ *
+ * @param linked_mod, if not NULL, address of pointer on linked modulators
+ *  list returned. NULL is returned in this pointer if linked
+ *  modulators doesn't exist in list_mod.
+ * @return
+ *  - the number of linked modulators if any valid linked path exists.
+ *  - 0 if no linked path exists.
+ *  - FLUID_FAILED if failed (memory error).
+ */
+int
+fluid_list_check_linked_mod(char *list_name, fluid_mod_t *list_mod,
+                            fluid_mod_t **linked_mod)
+{
+    int result;
+    /* path is a flags table state to register valid modulators belonging
+       to valid complete linked modulator paths */
+    unsigned char *path;
+    fluid_mod_t *mod;
+    int count; /* number of modulators in list_mod. */
+
+    /* count number of modulators in list_mod */
+    count = fluid_get_count_mod(list_mod);
+    if(!count)
+    { /* There are no modulators, no need to go further */
+        return 0;
+    }
+
+    /* path allocation */
+    path = FLUID_MALLOC (sizeof(*path) * count);
+    if(path == NULL)
+    {
+        return FLUID_FAILED;
+    }
+    /* initialize path: 
+	   - reset bits FLUID_PATH_VALID, FLUID_PATH_CURRENT
+	   - set bits FLUID_MOD_VALID
+	*/
+    FLUID_MEMSET(path, FLUID_MOD_VALID, count);
+
+    /* checks valid modulator sources (specs SF 2.01  7.4, 7.8, 8.2.1).*/
+    /* checks identic modulators in the list (specs SF 2.01  7.4, 7.8). */
+    if(list_name != NULL)
```

Comment:
Should the number of checks this function makes really depend on whether or not the first argument to this function is `NULL`? Wouldn't it be better to let this function always execute all checks?